### PR TITLE
download_sysext: check for download failure

### PR DESCRIPTION
**Pull request by dongsupark**
Define `DownloadFailed` status, and check for errors of HTTP client.
Check for HTTP status codes for immediate failures in `download_and_hash`, for now only `403 Forbidden` and `404 Not found`.

Fixes https://github.com/flatcar/ue-rs/issues/14.

### CODE REVIEW

File: src/download.rs [hunk 1]
```diff
@@ -14,11 +16,23 @@ where
...
+    ];
+
     #[rustfmt::skip]
     let mut res = client.get(url)
         .send()
         .await?;
 
+    // Return immediately on download failure on the client side.
+    let status = res.status();
+    if status.is_client_error() && failure_status.contains(&status) {
```
Comment (wrl) [inline src/download.rs:32]: probably better to just do a match statement for the `failure_status` checking – there's no real benefit to having the statuses in an array and also `contains()` is going to do a linear search through the array whereas a match statement wouldn't

File: src/bin/download_sysext.rs [hunk 2]
```diff
@@ -262,7 +270,10 @@ async fn main() -> Result<(), Box<dyn Error>> {
     for pkg in pkgs_to_dl.iter_mut() {
         pkg.check_download(&unverified_dir)?;
 
-        pkg.download(&unverified_dir, &client).await?;
+        match pkg.download(&unverified_dir, &client).await {
+            Ok(_) => (),
+            _ => return Err(format!("unable to download \"{}\"", pkg.name).into()),
+        };
```
Comment (wrl) [inline src/bin/download_sysext.rs:276]: matter of taste, but you could probably do this as something like

```rust
pkg.download(&unverified_dir, &client).await
    .map_err(|_| format!("unable to download \"{}\"", pkg.name).into())?;
```

**Review by wrl — CHANGES_REQUESTED** [review]

File: src/download.rs (hunk 1 above)
Comment (dongsupark) [reply src/download.rs]: Ok, updated as you suggested.

File: src/bin/download_sysext.rs (hunk 2 above)
Comment (dongsupark) [reply src/bin/download_sysext.rs:276]: Changing that, I am getting this:
```
error[E0283]: type annotations needed
   --> src/bin/download_sysext.rs:274:22
    |
274 |             .map_err(|_| format!("unable to download \"{}\"", pkg.name).into())?;
    |                      ^^^                                                       - type must be known at this point
```

For now, I left that unchanged.

Comment (dongsupark) [reply src/bin/download_sysext.rs:276]: In my other branch for anyhow migration, this kind of error mapping has been being already [done](https://github.com/flatcar/ue-rs/commit/f770af43d08e0a396266362a3acb5d1268c2ddf5#diff-42d8da6bd9394771306af4f16265896214311da363105863c47c2a9960a7eb82R79). 
Fortunately anyhow seems to allow that, without compile error.
Let's do the change in the anyhow migration PR, which I will create after merging the pending PRs. Ok?

**Review by wrl — APPROVED** [review]

File: src/download.rs [hunk 3]
```diff
@@ -19,6 +21,17 @@ where
         .send()
         .await?;
 
+    // Return immediately on download failure on the client side.
+    let status = res.status();
+    if status.is_client_error() {
```
Comment (pothos) [inline src/download.rs:26]:
```suggestion
    if status.is_client_error() || status.is_server_error() {
```

File: src/download.rs [hunk 4]
```diff
@@ -19,6 +21,17 @@ where
...
+        match status {
+            StatusCode::FORBIDDEN | StatusCode::NOT_FOUND => {
+                return Err(format!("cannnot fetch remotely with status code {:?}", status).into());
+            }
+            _ => (),
```
Comment (pothos) [inline src/download.rs:31]: We shouldn't continue here but also fail for now until we add retry logic.

File: src/download.rs (hunk 3 above)
Comment (pothos) [reply src/download.rs:26]: We should also either make use of https://docs.rs/reqwest/latest/reqwest/redirect/struct.Attempt.html#method.follow or check `is_redirection` and act accordingly.

File: src/download.rs (hunk 4 above)
Comment (dongsupark) [reply src/download.rs]: Done

File: src/download.rs (hunk 3 above)
Comment (dongsupark) [reply src/download.rs]: Updated to `!status.is_success()`.

Created a new issue for the redirect issue, https://github.com/flatcar/ue-rs/issues/27.
That looks out of scope of this PR. Probably needs more work, which I would like to add after we merge pending PRs including anyhow.

Comment (pothos) [reply src/download.rs]: Thanks!

**Review by pothos — APPROVED** [review]

**Pull request merged**